programs/survival/waystones.sc

```diff
@@ -0,0 +1,310 @@
+// [x,y,z] > name, dimension, icon
+global_waystones = {};
+// keep track of the waypoints each player has discovered.
+// uuid > [x,y,z](waystone pos) > [x,y,z](tp safe pos)
+global_waypoints = {};
+
+
+__on_start() -> (
+    _read_waystones();
+    file = read_file('waystone_settings', 'json');
+    if(file, global_settings = file);
+);
+__on_close() -> (
+    _write_waystones();
+    write_file('waystone_settings', 'json', global_settings);
+);
+
+_app_message(p, m) -> (
+    display_title(p, 'actionbar', format('wb '+m));
+);
+
+setting_read(key) -> (
+    print(player(), str('%s: %s', key, global_settings:key));
+);
+
+setting_command(val, key) -> (
+    global_settings:key = val;
+    print(player(), str('%s: %s', key, global_settings:key));
+);
+
+__on_player_places_block(p, item, hand, block)->(
+    if(item:0=='lodestone',
+        pos = pos(block);
+        floor = str(block(pos:0, pos:1 - 1, pos:2));
+        if(_has_required_structure(pos),
+            global_waystones:pos = {
+                'icon'->floor,
+                'dimension'->current_dimension(),
+                'name'->_get_name_from_nbt(item:2) || floor+' Waypoint' 
+            };
+            _app_message(p, str('Activated %s Waystone', global_waystones:pos:'name'));
+            _write_waystones();
+            sound( 'block.enchantment_table.use', pos );
+            sound( 'minecraft:block.amethyst_block.hit', pos );
+        ,
+            _app_message(p, str('Waystone structure requires %d %s to activate',
+                global_settings:'structure_size',global_settings:'structure_material'));
+            sound( 'block.copper.step', pos );
+        );
+    );
+);
+
+__on_player_breaks_block(p, block)->(
+    if(block~'lodestone',
+        pos = pos(block);
+        if(global_waystones:pos,
+            sound( 'minecraft:item.trident.thunder', pos );
+            _app_message(p, str('Removed %s Waystone', global_waystones:pos:'name'));
+            delete(global_waystones:pos); 
+        );
+    );
+);
+
+__on_player_right_clicks_block(p, item, hand, block, face, hitvec) -> (
+    if(block~'lodestone' && !p~'sneaking',
+        dimension = current_dimension();
+        if(global_settings:'enabled_dimensions'~dimension,
+            pos = pos(block);
+            
+            if(global_waystones:pos,
+                if( !_handle_name_tag(item, pos, p, hand) && hand=='mainhand',
+                    uuid = p~'uuid';
+                    _read_player_waypoints(uuid);
+                    _mark_player_waypoint(pos, p~'pos', uuid); 
+                    sound( 'minecraft:block.amethyst_block.hit', block );
+                    _open_waypoint_screen(p, pos, uuid);
+                    _write_player_waypoints(uuid);
+                );
+            ,
+                _app_message(p, str('Place the lodestone near a %d block %s structure to activate',
+                    global_settings:'structure_size',global_settings:'structure_material'));
```

Comment:
It can get a little confusing when you place the lodestone first, then the structure until you know it.

Reply:
Yeah, the script really should allow you to add the structure after the fact. perhaps it could check if you added a structure when you click in the stone. Ill look into that.

Reply:
it now attempts to activate a lodestones instead of rendering that message.